File: src/agentsfleetd/fleet/affinity.zig
//! fleet.runner_affinity — the per-fleet lease SLOT: the atomic claim, the
//! monotonic fencing source, and the sticky-routing hint, all on one row.
//!
//! `claim` is a single conditional UPSERT: it wins the fleet iff the slot is
//! free or its prior claim has expired (`leased_until < now`), bumping the
//! monotonic `fencing_seq` and recording the runner as the sticky hint. Exactly
//! one of N racing runners wins the row; losers get `.taken` and move on — and
//! crucially the claim precedes the event read, so a loser has consumed no
//! event (nothing is orphaned). `release` frees the slot at report, but is
//! token-guarded (`WHERE fencing_seq = token`) so a holder superseded by a
//! reclaim cannot free the current holder's slot; a dead runner never releases,
//! so its claim expires and another runner re-claims with a strictly higher
//! token. The report-time fence itself is a compare-and-swap in `service_report`
//! against this same `fencing_seq`.
//!
//! All functions run on a caller-supplied pooled connection (drained via
//! PgQuery / conn.exec).

const std = @import("std");
const sql = @import("sql.zig");
const clock = @import("common").clock;
const pg = @import("pg");
const PgQuery = @import("../db/pg_query.zig").PgQuery;
const id_format = @import("../types/id_format.zig");

/// The won claim: the new monotonic fencing token + the expiry the slot (and
/// the issued lease row) carry.
pub const Won = struct {
    token: u64,
    leased_until: i64,
};

/// Outcome of a claim attempt.
pub const Claim = union(enum) {
    /// The slot was won.
    won: Won,
    /// A live runner still holds the slot — try another fleet. No event read.
    taken,
};

/// Atomically claim the fleet's lease slot for `runner_id`, valid for
/// `ttl_ms`. Wins iff the slot is unclaimed or its prior claim has expired;
/// bumps the monotonic fencing token and records the sticky hint. Returns
/// `.taken` when a live runner still holds it.
///
/// The durable metering cursor is seeded `0`/now on a brand-new slot and is
/// deliberately ABSENT from the `ON CONFLICT` SET — so it is preserved across a
/// reclaim (the re-leased run meters forward from the dead holder's progress).
/// A fresh event resets it at lease issue; the renewal CTE advances it.
pub fn claim(
    conn: *pg.Conn,
    alloc: std.mem.Allocator,
    fleet_id: []const u8,
    runner_id: []const u8,
    ttl_ms: i64,
) !Claim {
    const affinity_id = try id_format.generateRunnerAffinityId(alloc);
    defer alloc.free(affinity_id);
    const now_ms = clock.nowMillis();
    const leased_until = now_ms + ttl_ms;
    var q = PgQuery.from(try conn.query(sql.CLAIM_AFFINITY_SLOT, .{ affinity_id, fleet_id, runner_id, leased_until, now_ms }));
    defer q.deinit();
    const row = try q.next() orelse return .taken;
    return .{ .won = .{ .token = @intCast(try row.get(i64, 0)), .leased_until = leased_until } };
}

/// Reset the per-fleet metering cursor to 0/now — called at FRESH lease issue
/// so a new event meters from zero even when the slot was reused from a prior
/// (completed) run whose cursor the claim's `ON CONFLICT` preserved. A reclaim
/// does NOT call this: the slot must keep the dead holder's progress so the
/// re-leased run meters forward from where it stopped. The renewal CTE reads
/// this cursor for each slice's Δ, so a stale value here would over-charge the
/// first renewal — hence the reset is fail-closed (a reset error fails lease
/// issue rather than risk an over-charge). `meter_slice_seq` resets too so the
/// new event's breakdown numbering restarts at 1.
pub fn resetCursor(conn: *pg.Conn, fleet_id: []const u8, now_ms: i64) !void {
    _ = conn.exec(sql.RESET_AFFINITY_METERS, .{ fleet_id, now_ms }) catch return error.AffinityCursorResetFailed;
}

/// Free the slot (report / abandoned no-work claim) so the fleet's next event
/// is claimable — but only when `token` still equals the live `fencing_seq`, so
/// a holder superseded by a reclaim cannot free the current holder's slot.
/// Idempotent: a no-op if the row is gone or the token has been bumped.
pub fn release(conn: *pg.Conn, fleet_id: []const u8, token: u64) !void {
    const now_ms = clock.nowMillis();
    _ = conn.exec(sql.RELEASE_AFFINITY_SLOT, .{ fleet_id, now_ms, @as(i64, @intCast(token)) }) catch return error.AffinityReleaseFailed;
}

File: rustd/crates/afd_fleet/src/lease/affinity.rs
//! `fleet.runner_affinity` — the per-fleet lease SLOT: the atomic claim, the
//! monotonic fencing source, and the sticky-routing hint, all on one row.
//!
//! `affinity.zig` is the original. What is ported is the GUARANTEE — exactly
//! one of N racing runners wins a fleet, and the winner carries a token that is
//! strictly greater than every token issued before it. Two things are
//! deliberately NOT ported. Zig answers with a `union(enum) { won, taken }`
//! because it has no nullable struct return, where here "no slot won" is an
//! absence and says so as [`Option::None`]. And Zig passes a `*pg.Conn` into
//! each function because the caller is the only thing that can own one — here
//! the verbs are methods on [`Leases`], which owns the pool and keeps it
//! `pub(crate)`, so nothing outside this crate can run a statement that is not
//! in [`crate::sql`].
//!
//! # Why the token is a type
//!
//! A lease row binds five `i64`s in a row — `event_created_at`,
//! `fencing_token`, `leased_until`, and two instants — and two of them
//! transposed compiles clean and writes a lease that can never be reported
//! against. [`Fence`] exists so that only the value a claim minted can reach
//! the column that fences reports, and §3 can demand one by type rather than by
//! comment.

use afd_core::clock::UnixMillis;
use afd_core::id::Uuid7;
use sqlx::Row as _;

use crate::error::{Result, query};
use crate::lease::sql;
use crate::lease::store::Leases;

/// Statement name, for the context a query failure carries.
const CONTEXT_CLAIM: &str = "affinity claim";

/// Statement name, for the context a query failure carries.
const CONTEXT_RELEASE: &str = "affinity release";

/// Statement name, for the context a query failure carries.
const CONTEXT_RESET: &str = "affinity meter reset";

/// The `fencing_seq` column, which is the one number that orders lease holders.
///
/// Monotonic per fleet and minted only by [`Leases::claim`]: every winning
/// claim bumps it, so a holder superseded by a reclaim carries a strictly
/// smaller value than the runner that displaced it. That is the whole basis of
/// Invariant 2, and the reason a report verifies its token inside the same
/// statement that flips the lease.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct Fence(i64);

impl Fence {
    /// The token as the column stores it.
    #[must_use]
    pub const fn as_i64(self) -> i64 {
        self.0
    }

    /// The token as the wire spells it.
    ///
    /// `fencing_seq` is a monotonic counter the claim statement starts at one
    /// and only ever increments, so no value this crate produces is negative.
    /// One edited out of band could be, and it saturates to ZERO rather than
    /// panicking or taking its magnitude: zero is below every token a claim
    /// can mint, so a corrupted row fences itself out. `unsigned_abs` alone
    /// would turn `-1` into `1`, which is a token another holder may legitimately
    /// hold — a plausible wrong answer, which is the worse failure.
    #[must_use]
    pub const fn as_u64(self) -> u64 {
        if self.0 < 0 { 0 } else { self.0.unsigned_abs() }
    }

    /// A token as read back from a row.
    ///
    /// `pub(crate)` on purpose: outside this crate a `Fence` can only come from
    /// a claim or from a row this crate read, never from arithmetic on an
    /// `i64` a caller happened to have.
    pub(crate) const fn from_i64(value: i64) -> Self {
        Self(value)
    }
}

/// A won claim: the new token, and the instant the slot — and the lease issued
/// against it — stays valid until.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Claimed {
    /// The monotonic token this claim minted.
    pub fence: Fence,
    /// When the claim lapses and the slot becomes winnable again.
    pub leased_until: UnixMillis,
}

impl Leases {
    /// Atomically claim `fleet_id`'s lease slot for `runner_id`, valid for
    /// `ttl_ms`.
    ///
    /// Wins iff the slot is unclaimed or its prior claim has expired, bumping
    /// the monotonic token and recording the sticky hint. Answers `None` when a
    /// live runner still holds it — an ordinary outcome the caller reads as
    /// "try the next candidate", not a failure.
    ///
    /// The claim PRECEDES the event read by design: a loser has consumed no
    /// event, so nothing is orphaned by losing.
    ///
    /// # Errors
    /// Reports a datastore that would not answer. A refused claim is
    /// `Ok(None)`.
    pub async fn claim(
        &self,
        fleet_id: &Uuid7,
        runner_id: &Uuid7,
        now: UnixMillis,
        ttl_ms: i64,
    ) -> Result<Option<Claimed>> {
        let leased_until = now.saturating_add_millis(ttl_ms);
        let mut connection = self.pool().acquire().await?;
        let won = sqlx::query(sql::lease::CLAIM_AFFINITY_SLOT)
            .bind(fleet_id.as_str())
            .bind(runner_id.as_str())
            .bind(leased_until.as_millis())
            .bind(now.as_millis())
            .fetch_optional(&mut *connection)
            .await
            .map_err(query(CONTEXT_CLAIM))?;

        // No row is the `.taken` verdict: the conditional UPSERT's WHERE clause
        // declined to update, which means someone else's claim is still live.
        let Some(row) = won else {
            return Ok(None);
        };
        let fence: i64 = row.try_get(0).map_err(query(CONTEXT_CLAIM))?;
        Ok(Some(Claimed {
            fence: Fence::from_i64(fence),
            leased_until,
        }))
    }

    /// Free the slot so the fleet's next event is claimable.
    ///
    /// Token-guarded: frees it only while `fence` is still the live token, so a
    /// holder superseded by a reclaim cannot free the CURRENT holder's slot and
    /// hand one fleet to two runners. Idempotent — a no-op when the row is gone
    /// or the token has moved on.
    ///
    /// Called on every post-claim path that does not issue a lease, so an
    /// abandoned claim costs one poll rather than a full TTL of silence on that
    /// fleet.
    ///
    /// # Errors
    /// Reports a datastore that would not answer.
    pub async fn release(&self, fleet_id: &Uuid7, fence: Fence, now: UnixMillis) -> Result<()> {
        let mut connection = self.pool().acquire().await?;
        sqlx::query(sql::lease::RELEASE_AFFINITY_SLOT)
            .bind(fleet_id.as_str())
            .bind(now.as_millis())
            .bind(fence.as_i64())
            .execute(&mut *connection)
            .await
            .map_err(query(CONTEXT_RELEASE))?;
        Ok(())
    }

    /// Reset the slot's metering cursor to zero at a FRESH lease issue.
    ///
    /// A reclaim must NOT call this: the slot has to keep the dead holder's
    /// progress so the re-leased run meters forward from where it stopped,
    /// which is exactly why the cursor is absent from the claim's `ON CONFLICT`
    /// SET.
    ///
    /// Fail-closed by contract — the caller treats an error here as a failed
    /// lease issue rather than a warning, because the renewal CTE reads this
    /// cursor for each slice's delta and a stale value would over-charge the
    /// first renewal.
    ///
    /// # Errors
    /// Reports a datastore that would not answer.
    pub async fn reset_meters(&self, fleet_id: &Uuid7, now: UnixMillis) -> Result<()> {
        let mut connection = self.pool().acquire().await?;
        sqlx::query(sql::lease::RESET_AFFINITY_METERS)
            .bind(fleet_id.as_str())
            .bind(now.as_millis())
            .execute(&mut *connection)
            .await
            .map_err(query(CONTEXT_RESET))?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A fence only ever moves through the column's own type.
    ///
    /// Cheap, and it is the property that makes the newtype worth its weight:
    /// what goes into the row is what came out of the claim.
    #[test]
    fn test_a_fence_round_trips_through_its_column_type() {
        let fence = Fence::from_i64(7);
        assert_eq!(fence.as_i64(), 7, "the token reaches the column unchanged");
    }

    /// Fences order, because that ordering IS the staleness test.
    ///
    /// §3 rejects a report whose token is behind the slot's current one, so an
    /// ordering that did not hold would be a stale writer admitted.
    #[test]
    fn test_a_later_fence_outranks_an_earlier_one() {
        assert!(
            Fence::from_i64(2) > Fence::from_i64(1),
            "a reclaim's token must outrank the holder it displaced"
        );
    }
}
